lib/functions.php

```diff
@@ -251,32 +254,32 @@ function build(array $parts): string
     $uri = '';
 
     $authority = '';
-    if (!empty($parts['host'])) {
+    if (isset($parts['host'])) {
         $authority = $parts['host'];
-        if (!empty($parts['user'])) {
+        if (isset($parts['user'])) {
```

Comment:
These changes from `!empty()` to `isset()` now allow the value to be "0".
I added unit test cases for ways that I could see that the value of these things might be "reasonably" passed in as the value "0". For example, a user "name" might be the string "0"! Or the query string could be just `?0` (very edge case - an application that just sends a number in the query string). Or the fragment could be `#0`. All of those unit test cases fail with the `!empty()` check, and pass with `isset()`

Reply:
Note: we need to use `isset()` because the `$parts` array passed to `build()` may not have all the component keys set.

In other places we could compare directly to `null` because `resolve()`, `normalize()` etc. we guarantee that all the URI components are actually set in the array (and "missing" components are set to value `null`